### ISSUE TITLE
Disable DiscreteNorm by default for certain plots

### ISSUE THREAD

**Issue by naveenaero** [lukelbd/proplot#233]
### Description

Proplot's imshow function does not behave in the same way as it would behave it used directly from matplotlib. If I plot an image, instead of seeing a smooth image, I see contour levels just like you would expect from a matplotliib contourf plot. 

### Steps to reproduce and Equivalent steps in matplotlib

Code which generates 3 plots of  f = exp(-X**2-Y**2), 
1. proplot imshow (actual behaviour)
2. matplotlib imshow (expected behaviour)
3. matplotlib contourf (similar to actual behaviour but in matplotlib, I think somewhere internally the proplot imshow wrapper just call matplotlib contourf)

import proplot as plot
import numpy as np
import matplotlib.pyplot as plt

x = np.linspace(0,1,100)
[X,Y] = np.meshgrid(x,x)
f = np.exp(-X\*\*2-Y\*\*2)
fig1, axs = plot.subplots()
axs.format(title="proplot imshhow")
axs.imshow(f)
plt.show()
fig2 = plt.figure()
plt.imshow(f)
plt.title("matplotlib imshow")
plt.show()
fig3 = plt.figure()
plt.contourf(f,levels=11)
plt.title('matplotlib contourf')
plt.show()

### Proplot version

0.6.4

**Comment by lukelbd** [comment]
Hi @naveenaero, I think this qualifies as "not a bug, but a feature." Proplot tries to consistently divide colormaps into distinct levels with [`DiscreteNorm`](https://proplot.readthedocs.io/en/latest/api/proplot.colors.DiscreteNorm.html) wherever colormaps are used. The rationale is that it's easier to visually relate data values to the distinct levels. However if you do need a "smooth-looking" colormap you can simply use a very large number of levels:

```python
x = np.linspace(0, 1, 100)
[X, Y] = np.meshgrid(x, x)
Z = np.exp(-X**2-Y**2)
fig, ax = plot.subplots()
ax.imshow(Z, N=500)
```

![example](https://user-images.githubusercontent.com/19657652/123954288-743cd480-d965-11eb-8c03-3d073f5b19a7.png)

**Comment by lukelbd** [comment]
On reflection I think I'll re-open this... I see that this is frequently confusing/annoying for users so probably the best approach is to disable `DiscreteNorm` by default for certain types of plots (probably `imshow`, `matshow`, and maybe `scatter`) and allow explicitly turning it off and on with a keyword argument like `discrete=False` or `discretenorm=False`.

**Comment by lukelbd** [comment]
I think a good paradigm might be to add the rc setting `rc['image.discretenorm']` with the following options:

* If `None`, enable or disable the `DiscreteNorm` depending on plot type (e.g. off for `imshow`, on for `contourf`).
* If `False`, always disable `DiscreteNorm`.
* If `True`, always enable `DiscreteNorm`.

**Comment by lukelbd** [comment]
This is now fixed! See 85cec7b

Now the plotting commands wrapped by [`apply_cmap`](https://proplot.readthedocs.io/en/latest/api/proplot.axes.apply_cmap.html#proplot.axes.apply_cmap) have a boolean `discrete` option that enables/disables application of `DiscreteNorm`. By default, proplot uses the new rc setting `rc['image.discrete']`. When `discrete` is `None`, proplot turns off discrete norm for `imshow`, `matshow`, `hexbin`, `hist2d`, and `spy` plots by default.

So, to get a densely-sampled "smooth" colormap in any plot, you now have two options:

1. Use `levels=large_number` (probably slower).
2. Use `discrete=False`.

Here's an example showing the new default behavior. Note `imshow` automatically gets a "smooth" colormap.


```python
N = 15
state = np.random.RandomState(51423)
data = state.rand(N, N).cumsum(axis=0)

# Default levels
fig, axs = pplt.subplots(ncols=2)
ax = axs[0]
ax.imshow(data, colorbar='b')
ax.set_title('Imshow plot')
ax = axs[1]
ax.pcolormesh(data, colorbar='b')
ax.set_title('Pcolor plot')
```

![test](https://user-images.githubusercontent.com/19657652/125208676-2ec7b380-e251-11eb-87fe-7267fffc4d2d.png)

**Comment by lukelbd** [comment]
Also here's an example I added to the website:

```python
import proplot as pplt
import numpy as np

# Sample data
state = np.random.RandomState(51423)
data = (state.normal(0, 1, size=(33, 33))).cumsum(axis=0).cumsum(axis=1)

# Figure
fig, axs = pplt.subplots([[1, 1, 2, 2], [0, 3, 3, 0]], ref=3, refwidth=2.3)
axs.format(yformatter='none', suptitle='Distinct vs. smooth colormap levels')

# Pcolor with DivergingNorm
axs[0].pcolor(data, cmap='spectral_r', norm='div', colorbar='l')
axs[0].set_title('Pcolor plot\nDiscreteNorm enabled (default)')
axs[1].pcolor(data, discrete=False, cmap='spectral_r', norm='div', colorbar='r')
axs[1].set_title('Pcolor plot\nDiscreteNorm disabled')

# Imshow
m = axs[2].imshow(data, cmap='roma', colorbar='b')
axs[2].format(title='Imshow plot\nDiscreteNorm disabled (default)', yformatter='auto')
```

![tmp](https://user-images.githubusercontent.com/19657652/125210510-80763b00-e25d-11eb-88c1-da0c8ae3724a.png)
